File: ast.java
// COSC57 AST class for Java
// By Ben Kallus

import java.util.ArrayList;
import java.util.Optional;


// Relational operators
enum ASTROpType {
    LT,  // <
    GT,  // >
    LEQ,  // <=
    GEQ,  // >=
    EQ,  // ==
    NEQ;  // !=
}

// Binary operators
enum ASTBOpType {
    ADD,  // +
    SUB,  // -
    DIV,  // /
    MUL;  // *
}

// Unary operators
enum ASTUOpType {
    POS,  // +
    NEG;  // -
}

// The types in miniC
enum ASTDataType {
    VOID_T,
    INT_T,
    FLOAT_T,
    BOOL_T;
}


// Abstract class from which all node classes in the AST derive.
abstract class ASTNode {
    public abstract void print(int n);
    public void print() {
        print(0);
    }
    String INDENT_STR = "|   ";
}


// Abstract class from which all statements derive.
// A statement is (informally) anything that could sit on its own in a block.
abstract class ASTStmtNode extends ASTNode {}


// Abstract class from which variable and function declarations derive.
abstract class ASTDeclNode extends ASTStmtNode {}


// Abstract class from which all expressions derive.
// An expression is a term, a function call, or expressions combined with appropriate operators.
abstract class ASTExprNode extends ASTStmtNode {}


// An assignment (eg. "a = 1;")
class ASTAsgnNode extends ASTStmtNode {
    public String lhs;
    public ASTExprNode rhs;

    public ASTAsgnNode(String lhs_p, ASTExprNode rhs_p) {
        // The left-hand side of the assignment
        lhs = lhs_p;
        // The right-hand side of the assignment
        rhs = rhs_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sAsgn. %s =%n", INDENT_STR.repeat(indentation_level), lhs);
        rhs.print(indentation_level + 1);
    }
}


// A block (eg. "{ int a; a = 10 + b; exit(1); return will_never_get_here; }")
class ASTBlockNode extends ASTStmtNode {
    public ArrayList<ASTStmtNode> stmt_list;

    public ASTBlockNode(ArrayList<ASTStmtNode> stmt_list_p) {
        // The statements that make up the block
        stmt_list = stmt_list_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sBlock.%n", INDENT_STR.repeat(indentation_level));
        for (ASTStmtNode stmt : stmt_list) {
            stmt.print(indentation_level + 1);
        }
    }
}


// A function definition (eg. "int the_identity_fn(int n) { return n; }")
class ASTFuncDefNode extends ASTNode {
    public ASTDataType return_type;
    public String name;
    public ASTDataType param_type;
    public String param_name;
    public ASTBlockNode body;

    public ASTFuncDefNode(ASTDataType return_type_p, String name_p, ASTDataType param_type_p, String param_name_p, ASTBlockNode body_p) {
        // The return type of the function
        return_type = return_type_p;
        // The name of the function
        name = name_p;
        // The parameter type of the function
        param_type = param_type_p;
        // The parameter name of the function
        param_name = param_name_p;  // If there is no parameter name, just use ""
        // The function body
        body = body_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sFuncDef. %s %s(%s %s)%n", INDENT_STR.repeat(indentation_level), return_type, name, param_type, param_name);
        body.print(indentation_level + 1);
    }
}


// A root node of an AST. You should have one of these per compilation unit.
class ASTRootNode extends ASTNode {
    public ArrayList<ASTDeclNode> decls;
    public ArrayList<ASTAsgnNode> asgns;
    public ArrayList<ASTFuncDefNode> funcs;

    public ASTRootNode(ArrayList<ASTDeclNode> decls_p, ArrayList<ASTAsgnNode> asgns_p, ArrayList<ASTFuncDefNode> funcs_p) {
        // List of function and variable declarations.
        // Must be at the top of the file.
        decls = decls_p;
        // List of variable assignments for global variables.
        // Must be after declarations and before functions in the file.
        asgns = asgns_p;
        // List of function definitions.
        // Must be at the end of the file.
        funcs = funcs_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sRoot.%n", INDENT_STR.repeat(indentation_level));
        if (asgns.isEmpty()) {
            System.out.printf("%sAssignments:%n", INDENT_STR.repeat(indentation_level));
        }
        for (ASTAsgnNode asgn : asgns) {
            asgn.print(indentation_level + 1);
        }
        if (decls.isEmpty()) {
            System.out.printf("%sDeclarations:%n", INDENT_STR.repeat(indentation_level));
        }
        for (ASTDeclNode decl : decls) {
            decl.print(indentation_level + 1);
        }
        if (funcs.isEmpty()) {
            System.out.printf("%sFunctions:%n", INDENT_STR.repeat(indentation_level));
        }
        for (ASTFuncDefNode func : funcs) {
            func.print(indentation_level + 1);
        }
    }
}


// A variable declaration. (eg. "int a;")
class ASTVarDeclNode extends ASTDeclNode {
    public Boolean is_extern;
    public ASTDataType data_type;
    public String name;

    public ASTVarDeclNode(Boolean is_extern_p, ASTDataType data_type_p, String name_p) {
        // Whether the variable is extern
        is_extern = is_extern_p;
        // The type of the variable
        data_type = data_type_p;
        // The name of the variable
        name = name_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sVarDecl. %s%s %s%n", INDENT_STR.repeat(indentation_level), is_extern ? "extern " : "", data_type, name);
    }
}


// A function declaration (eg. "int main(void);" or "extern int square(int);")
class ASTFuncDeclNode extends ASTDeclNode {
    public Boolean is_extern;
    public ASTDataType return_type;
    public String name;
    public ASTDataType param_type;

    public ASTFuncDeclNode(Boolean is_extern_p, ASTDataType return_type_p, String name_p, ASTDataType param_type_p) {
        // Whether the function is extern
        is_extern = is_extern_p;
        // The return type of the function
        return_type = return_type_p;
        // The name of the function
        name = name_p;
        // The type of the parameter to the function
        param_type = param_type_p;
        // Note that parameter names don't matter in C function declarations because all arguments are positional.
        // Thus, we don't store that info in the AST.
    }

    public void print(int indentation_level) {
        System.out.printf("%sFuncDecl. %s %s %s(%s)%n", INDENT_STR.repeat(indentation_level), is_extern ? "extern " : "", return_type, name, param_type);
    }
}


// A variable term (eg. the "a" in "b = 10 + a * 2.1;")
class ASTVarNode extends ASTExprNode {
    public String name;

    public ASTVarNode(String name_p) {
        // The variable's name
        name = name_p;
        // Note that we don't know the variable's type here.
        // We'll get to that when we decorate the AST.
    }

    public void print(int indentation_level) {
        System.out.printf("%sVar. %s%n", INDENT_STR.repeat(indentation_level), name);
    }
}


// An integer constant term (eg. the "10" in "b = 10 + a * 2.1;")
class ASTIntLiteralNode extends ASTExprNode {
    public int value;

    public ASTIntLiteralNode(int value_p) {
        // The constant's value
        value = value_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sIntLiteral. %s%n", INDENT_STR.repeat(indentation_level), value);
    }
}


// A floating point constant term (eg. the "2.1" in "b = 10 + a * 2.1;")
class ASTFloatLiteralNode extends ASTExprNode {
    public float value;

    public ASTFloatLiteralNode(float value_p) {
        // The constant's value
        value = value_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sFloatLiteral. %s%n", INDENT_STR.repeat(indentation_level), value);
    }
}


// A boolean constant term (eg. "false")
class ASTBoolLiteralNode extends ASTExprNode {
    public Boolean value;

    public ASTBoolLiteralNode(Boolean value_p) {
        // The constant's value
        value = value_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sBoolLiteral. %s%n", INDENT_STR.repeat(indentation_level), value);
    }
}


// A relational expression (eg. "10 > 2")
class ASTRExprNode extends ASTExprNode {
    public ASTExprNode lhs;
    public ASTExprNode rhs;
    public ASTROpType op;

    public ASTRExprNode(ASTExprNode lhs_p, ASTExprNode rhs_p, ASTROpType op_p) {
        // The left-hand side of the relational operator
        lhs = lhs_p;
        // The right-hand side of the relational operator
        rhs = rhs_p;
        // The relational operator
        op = op_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sRExpr. %s%n", INDENT_STR.repeat(indentation_level), op);
        System.out.printf("%s(RExpr) LHS =%n", INDENT_STR.repeat(indentation_level));
        lhs.print(indentation_level + 1);
        System.out.printf("%s(RExpr) RHS =%n", INDENT_STR.repeat(indentation_level));
        rhs.print(indentation_level + 1);
    }
}


// A binary expression (eg. "10 + 2")
class ASTBExprNode extends ASTExprNode {
    public ASTExprNode lhs;
    public ASTExprNode rhs;
    public ASTBOpType op;

    public ASTBExprNode(ASTExprNode lhs_p, ASTExprNode rhs_p, ASTBOpType op_p) {
        // The left-hand side of the binary operator
        lhs = lhs_p;
        // The right-hand side of the binary operator
        rhs = rhs_p;
        // The binary operator
        op = op_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sBExpr. %s%n", INDENT_STR.repeat(indentation_level), op);
        System.out.printf("%s(BExpr) LHS =%n", INDENT_STR.repeat(indentation_level));
        lhs.print(indentation_level + 1);
        System.out.printf("%s(BExpr) RHS =%n", INDENT_STR.repeat(indentation_level));
        rhs.print(indentation_level + 1);
    }
}


// A unary expression (eg. "-10")
class ASTUExprNode extends ASTExprNode {
    public ASTExprNode expr;
    public ASTUOpType op;

    public ASTUExprNode(ASTExprNode expr_p, ASTUOpType op_p) {
        // The expression to which the unary operation applies
        expr = expr_p;
        // The unary operator
        op = op_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sUExpr. %s, expression =%n", INDENT_STR.repeat(indentation_level), op);
        expr.print(indentation_level + 1);
    }
}


// A function call (eg. "set_coolness_factor(100);")
class ASTCallNode extends ASTExprNode {
    public String name;
    public Optional<ASTExprNode> param;

    public ASTCallNode(String name_p, Optional<ASTExprNode> param_p) {
        // The name of the function being called
        name = name_p;
        // The parameter to the function
        param = param_p;  // Remember this is optional
    }

    public void print(int indentation_level) {
        System.out.printf("%sCall. %s%s%n", INDENT_STR.repeat(indentation_level), name, param.isPresent() ? ", parameter =" : "");
        if (param.isPresent()) {
            param.get().print(indentation_level + 1);
        }
    }
}


// A return statement (eg. "return 10;")
class ASTRetNode extends ASTStmtNode {
    public Optional<ASTExprNode> expr;

    public ASTRetNode(Optional<ASTExprNode> expr_p) {
        // The thing being returned
        expr = expr_p;  // Remember this is optional
    }

    public void print(int indentation_level) {
        System.out.printf("%sRet.%s%n", INDENT_STR.repeat(indentation_level), expr.isPresent() ? " expression =" : "");
        if (expr.isPresent()) {
            expr.get().print(indentation_level + 1);
        }
    }
}


// A while statement (eg. "while (1 == 1) { 1; }")
class ASTWhileNode extends ASTStmtNode {
    public ASTExprNode cond;
    public ASTBlockNode body;

    public ASTWhileNode(ASTExprNode cond_p, ASTBlockNode body_p) {
        // The loop condition
        cond = cond_p;
        // The loop body
        body = body_p;
    }

    public void print(int indentation_level) {
        System.out.printf("%sWhile.%n", INDENT_STR.repeat(indentation_level));
        System.out.printf("%s(While) condition =%n", INDENT_STR.repeat(indentation_level));
        cond.print(indentation_level + 1);
        System.out.printf("%s(While) body =%n", INDENT_STR.repeat(indentation_level));
        body.print(indentation_level + 1);
    }
}


// An if statement (eg. "if (1 != 1) { return BAD_NEWS; } else { return ALL_IS_WELL; }")
class ASTIfNode extends ASTStmtNode {
    public ASTExprNode cond;
    public ASTBlockNode if_body;
    public ASTBlockNode else_body;

    public ASTIfNode(ASTExprNode cond_p, ASTBlockNode if_body_p, Optional<ASTBlockNode> else_body_p) {
        // The if condition
        cond = cond_p;
        // The if body
        if_body = if_body_p;
        // The else body
        else_body = else_body_p.isPresent() ? else_body_p.get() : new ASTBlockNode(new ArrayList<ASTStmtNode>());
    }

    public void print(int indentation_level) {
        System.out.printf("%sIf.%n", INDENT_STR.repeat(indentation_level));
        System.out.printf("%s(If) condition =%n", INDENT_STR.repeat(indentation_level));
        cond.print(indentation_level + 1);
        System.out.printf("%s(If) if_body =%n", INDENT_STR.repeat(indentation_level));
        if_body.print(indentation_level + 1);
        System.out.printf("%s(If) else_body =%n", INDENT_STR.repeat(indentation_level));
        else_body.print(indentation_level + 1);
    }
}


// An empty statement (eg. ";")
// (This could just be ASTStmtNode, but I'd rather have that be an abstract class)
class ASTEmptyStmtNode extends ASTStmtNode {
    public ASTEmptyStmtNode() {}

    public void print(int indentation_level) {
        System.out.printf("%sEmptyStmt.%n", INDENT_STR.repeat(indentation_level));
    }
}

File: AST.py
### COSC57 AST class for Python
### By Ben Kallus
### Fully type checked by mypy; formatted with black.
### If you are cool, you will run your code through mypy and black when you use this library.
#
# 19 Apr 2022  bjk  Creation
# 24 Apr 2022  jpb  Released v1 for class
#

from enum import Enum
from typing import List, Optional

INDENT_STR = "|   "

# Enum's __str__ will spit out its class name by default.
# This class fixes that problem.
class MyEnum(Enum):
    def __str__(self):
        return self._name_


# Relational operators
class ASTROpType(MyEnum):
    LT = 0  # <
    GT = 1  # >
    LEQ = 2  # <=
    GEQ = 3  # >=
    EQ = 4  # ==
    NEQ = 5  # !=


# Binary operators
class ASTBOpType(MyEnum):
    ADD = 0  # +
    SUB = 1  # -
    DIV = 2  # /
    MUL = 3  # *


# Unary operators
class ASTUOpType(MyEnum):
    POS = 0  # +
    NEG = 1  # -


# The types in miniC
class ASTDataType(MyEnum):
    VOID_T = 0
    INT_T = 1
    FLOAT_T = 2
    BOOL_T = 3


# Abstract class from which all node classes in the AST derive.
class ASTNode:
    def print(self, indentation_level: int = 0):
        assert False


# Abstract class from which all statements derive.
# A statement is (informally) anything that could sit on its own in a block.
class ASTStmtNode(ASTNode):
    pass


# Abstract class from which variable and function declarations derive.
class ASTDeclNode(ASTStmtNode):
    pass


# Abstract class from which all expressions derive.
# An expression is a term, a function call, or expressions combined with appropriate operators.
class ASTExprNode(ASTStmtNode):
    pass


# An assignment (eg. "a = 1;")
class ASTAsgnNode(ASTStmtNode):
    def __init__(self, lhs: str, rhs: ASTExprNode) -> None:
        # The left-hand side of the assignment
        self.lhs = lhs
        # The right-hand side of the assignment
        self.rhs = rhs

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}Asgn. {self.lhs} =")
        self.rhs.print(indentation_level + 1)


# A block (eg. "{ int a; a = 10 + b; exit(1); return will_never_get_here; }")
class ASTBlockNode(ASTStmtNode):
    def __init__(self, stmt_list: List[ASTStmtNode]) -> None:
        # The statements that make up the block
        self.stmt_list = stmt_list

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}Block.")
        for stmt in self.stmt_list:
            stmt.print(indentation_level + 1)


# A function definition (eg. "int the_identity_fn(int n) { return n; }")
class ASTFuncDefNode(ASTNode):
    def __init__(
        self,
        return_type: ASTDataType,
        name: str,
        param_type: ASTDataType,
        param_name: str,
        body: ASTBlockNode,
    ) -> None:
        # The return type of the function
        self.return_type = return_type
        # The name of the function
        self.name = name
        # The parameter type of the function
        self.param_type = param_type
        # The parameter name of the function
        self.param_name = param_name  # If there is no parameter name, just use ""
        # The function body
        self.body = body

    def print(self, indentation_level: int = 0) -> None:
        print(
            f"{INDENT_STR * indentation_level}FuncDef. {self.return_type} {self.name}({self.param_type} {self.param_name})"
        )
        self.body.print(indentation_level + 1)


# A root node of an AST. You should have one of these per compilation unit.
class ASTRootNode(ASTNode):
    def __init__(
        self,
        decls: List[ASTDeclNode],
        asgns: List[ASTAsgnNode],
        funcs: List[ASTFuncDefNode],
    ) -> None:
        # List of function and variable declarations.
        # Must be at the top of the file.
        self.decls = decls
        # List of variable assignments for global variables.
        # Must be after declarations and before functions in the file.
        self.asgns = asgns
        # List of function definitions.
        # Must be at the end of the file.
        self.funcs = funcs

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}Root.")
        if self.asgns != []:
            print(f"{INDENT_STR * indentation_level}Assignments:")
        for asgn in self.asgns:
            asgn.print(indentation_level + 1)
        if self.decls != []:
            print(f"{INDENT_STR * indentation_level}Declarations:")
        for decl in self.decls:
            decl.print(indentation_level + 1)
        if self.funcs != []:
            print(f"{INDENT_STR * indentation_level}Functions:")
        for func in self.funcs:
            func.print(indentation_level + 1)


# A variable declaration. (eg. "int a;")
class ASTVarDeclNode(ASTDeclNode):
    def __init__(self, is_extern: bool, data_type: ASTDataType, name: str) -> None:
        # Whether the variable is extern
        self.is_extern = is_extern
        # The type of the variable
        self.data_type = data_type
        # The name of the variable
        self.name = name

    def print(self, indentation_level: int = 0) -> None:
        print(
            f"{INDENT_STR * indentation_level}VarDecl. {'extern ' * self.is_extern}{self.data_type} {self.name}"
        )


# A function declaration (eg. "int main(void);" or "extern int square(int);")
class ASTFuncDeclNode(ASTDeclNode):
    def __init__(
        self,
        is_extern: bool,
        return_type: ASTDataType,
        name: str,
        param_type: ASTDataType,
    ) -> None:
        # Whether the function is extern
        self.is_extern = is_extern
        # The return type of the function
        self.return_type = return_type
        # The name of the function
        self.name = name
        # The type of the parameter to the function
        self.param_type = param_type
        # Note that parameter names don't matter in C function declarations because all arguments are positional.
        # Thus, we don't store that info in the AST.

    def print(self, indentation_level: int = 0) -> None:
        print(
            f"{INDENT_STR * indentation_level}FuncDecl. {'extern' * self.is_extern} {self.return_type} {self.name}({self.param_type})"
        )


# A variable term (eg. the "a" in "b = 10 + a * 2.1;")
class ASTVarNode(ASTExprNode):
    def __init__(self, name: str) -> None:
        # The variable's name
        self.name = name
        # Note that we don't know the variable's type here.
        # We'll get to that when we decorate the AST.

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}Var. {self.name}")


# An integer constant term (eg. the "10" in "b = 10 + a * 2.1;")
class ASTIntLiteralNode(ASTExprNode):
    def __init__(self, value: int) -> None:
        # The constant's value
        self.value = value

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}IntLiteral. {self.value}")


# A floating point constant term (eg. the "2.1" in "b = 10 + a * 2.1;")
class ASTFloatLiteralNode(ASTExprNode):
    def __init__(self, value: float) -> None:
        # The constant's value
        self.value = value

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}FloatLiteral. {self.value}")


# A boolean constant term (eg. "false")
class ASTBoolLiteralNode(ASTExprNode):
    def __init__(self, value: bool) -> None:
        # The constant's value
        self.value = value

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}BoolLiteral. {self.value}")


# A relational expression (eg. "10 > 2")
class ASTRExprNode(ASTExprNode):
    def __init__(self, lhs: ASTExprNode, rhs: ASTExprNode, op: ASTROpType) -> None:
        # The left-hand side of the relational operator
        self.lhs = lhs
        # The right-hand side of the relational operator
        self.rhs = rhs
        # The relational operator
        self.op = op

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}RExpr. {self.op}")
        print(f"{INDENT_STR * indentation_level}(RExpr) LHS =")
        self.lhs.print(indentation_level + 1)
        print(f"{INDENT_STR * indentation_level}(RExpr) RHS =")
        self.rhs.print(indentation_level + 1)


# A binary expression (eg. "10 + 2")
class ASTBExprNode(ASTExprNode):
    def __init__(self, lhs: ASTExprNode, rhs: ASTExprNode, op: ASTBOpType) -> None:
        # The left-hand side of the binary operator
        self.lhs = lhs
        # The right-hand side of the binary operator
        self.rhs = rhs
        # The binary operator
        self.op = op

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}BExpr. {self.op}")
        print(f"{INDENT_STR * indentation_level}(BExpr) LHS =")
        self.lhs.print(indentation_level + 1)
        print(f"{INDENT_STR * indentation_level}(BExpr) RHS =")
        self.rhs.print(indentation_level + 1)


# A unary expression (eg. "-10")
class ASTUExprNode(ASTExprNode):
    def __init__(self, expr: ASTExprNode, op: ASTUOpType) -> None:
        # The expression to which the unary operation applies
        self.expr = expr
        # The unary operator
        self.op = op

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}UExpr. {self.op}, expression =")
        self.expr.print(indentation_level + 1)


# A function call (eg. "set_coolness_factor(100);")
class ASTCallNode(ASTExprNode):
    def __init__(self, name: str, param: Optional[ASTExprNode] = None) -> None:
        # The name of the function being called
        self.name = name
        # The parameter to the function
        self.param = param  # Could be None

    def print(self, indentation_level: int = 0) -> None:
        print(
            f"{INDENT_STR * indentation_level}Call. {self.name}{', parameter =' * (self.param is not None)}"
        )
        if self.param is not None:
            self.param.print(indentation_level + 1)


# A return statement (eg. "return 10;")
class ASTRetNode(ASTStmtNode):
    def __init__(self, expr: Optional[ASTExprNode] = None) -> None:
        # The thing being returned
        self.expr = expr  # Could be None

    def print(self, indentation_level: int = 0) -> None:
        print(
            f"{INDENT_STR * indentation_level}Ret.{' expression =' * (self.expr is not None)}"
        )
        if self.expr is not None:
            self.expr.print(indentation_level + 1)


# A while statement (eg. "while (1 == 1) { 1; }")
class ASTWhileNode(ASTStmtNode):
    def __init__(self, cond: ASTExprNode, body: ASTBlockNode) -> None:
        # The loop condition
        self.cond = cond
        # The loop body
        self.body = body

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}While.")
        print(f"{INDENT_STR * indentation_level}(While) condition =")
        self.cond.print(indentation_level + 1)
        print(f"{INDENT_STR * indentation_level}(While) body =")
        self.body.print(indentation_level + 1)


# An if statement (eg. "if (1 != 1) { return BAD_NEWS; } else { return ALL_IS_WELL; }")
class ASTIfNode(ASTStmtNode):
    def __init__(
        self,
        cond: ASTExprNode,
        if_body: ASTBlockNode,
        else_body: Optional[ASTBlockNode] = None,
    ) -> None:
        # The if condition
        self.cond = cond
        # The if body
        self.if_body = if_body
        # The else body
        self.else_body = ASTBlockNode(stmt_list=[]) if else_body is None else else_body

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}If.")
        print(f"{INDENT_STR * indentation_level}(If) condition =")
        self.cond.print(indentation_level + 1)
        print(f"{INDENT_STR * indentation_level}(If) if_body =")
        self.if_body.print(indentation_level + 1)
        print(f"{INDENT_STR * indentation_level}(If) else_body =")
        self.else_body.print(indentation_level + 1)


# An empty statement (eg. ";")
# (This could just be ASTStmtNode, but I'd rather have that be an abstract class)
class ASTEmptyStmtNode(ASTStmtNode):
    def __init__(self) -> None:
        pass

    def print(self, indentation_level: int = 0) -> None:
        print(f"{INDENT_STR * indentation_level}EmptyStmt.")


if __name__ == "__main__":
    print("This file isn't meant to be run on its own.")
